Fixes #4764 - HTTP2 Jetty Server does not send back content-length.

sbordet opened the pull request:
Sending Content-Length header if known at the time of sending the
response headers.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

Closes #4764.

sbordet committed:
Fixes #4764 - HTTP2 Jetty Server does not send back content-length.

Sending Content-Length header if known at the time of sending the
response headers.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java:
```diff
@@ -110,6 +111,13 @@ public void send(MetaData.Response info, boolean isHeadRequest, ByteBuffer conte
+                    {
+                        HttpFields responseHeaders = info.getFields();
+                        if (!responseHeaders.contains(HttpHeader.CONTENT_LENGTH))
+                            responseHeaders.put(HttpHeader.CONTENT_LENGTH, String.valueOf(BufferUtil.length(content)));
```
Does it have to be put into the `HttpFields` instance?  The HTTP/1 does this in the generator, so that as it is committing the headers it looks for an existing content-length header and if one is not found, but the length is known, then the content length header is generated directly into the output buffer without creating a `HttpField` instance or modifying the info.

The h2 approach here still results in different behaviour because the HttpFields class will be different after commit.

sbordet replied:
@gregw HTTP/2 works with frames, so if I can't put the `Content-Length` inside the `HttpFields` I would need an additional field `HeadersFrame.contentLength` which is doable but feels weird.

Also, it's not clear what should be done in case of conflicts with the application.
For example, if the application specified a different `Content-Length` value?

sbordet committed:
Fixes #4764 - HTTP2 Jetty Server does not send back content-length.

Updates after review.
Now the Content-Length header is generated by HpackEncoder based on
MetaData.contentLength, so that the MetaData.HttpFields are not modified.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-9.4.x' into 'jetty-9.4.x-4764-http2_content_length'.

sbordet committed:
Fixes #4764 - HTTP2 Jetty Server does not send back content-length.

Fixed InterleavingTest that was using the wrong MetaData.Response constructor.
Fixed handling of HEAD methods in HttpTransportOverHTTP2.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
